### ISSUE TITLE
FunctionManager Does Not Include Arguments of User Defined Functions.

### ISSUE THREAD

**Issue by H0r53** [NationalSecurityAgency/ghidra#4471]
I used a variant of [this script](https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Features/Base/ghidra_scripts/FindUndefinedFunctionsScript.java) to define all undefined functions within a binary I am analyzing. After which, I'm able to include local functions in the results of the following:

```
FunctionManager fm = currentProgram.getFunctionManager();
FunctionIterator funcItr = fm.getFunctions(true);

for (Function f : funcItr) {
	FunctionSignature sig = f.getSignature();			
	String prototypeStr = sig.getPrototypeString();
        println(prototypeStr);
}
```

However, I'm finding that all of the newly defined local functions are considered to have 0 arguments when iterating through them using the code above. For example, all of the previously defined `FUN_xxxxx` functions return 0 when I do `f.getSignature().getArguments().length`.

Here is a more concrete example. The following function prototype displays within Ghidra's Decompiler view.

```
void FUN_02d2d23c(int param_1,undefined8 param_2,uint param_3,undefined8 param_4,undefined2 param_5)
```

Yet, when I access this function through `currentProgram.getFunctionManager().getFunctions(true)` the prototype is listed as the following:

```
undefined FUN_02d2d23c(void)
```

Why do the signatures differ? Ghidra's Decompiler view is showing the correct prototype. Yet since FunctionManager always considers these functions to have zero arguments, I'm unable to leverage FunctionManager in my scripts to do more interesting things such as "Find all functions with a matching signature: (int arg1, long *arg2, void *arg3)"

**Comment by ghidracadabra** [comment]
```f.getSignature()``` returns the signature that's been committed to the program database for that function.  If a signature has been committed for ```f```, the decompiler will use that signature.  Otherwise it will attempt to infer the signature for ```f```.  Creating a function does not commit a signature.  You can commit signatures manually with the ```Commit Params/Return``` action in the decompiler or programmatically via the class ```HighFunctionDBUtils```.  

Additional information can be found in the "Program Annotations Affecting the Decompiler" section of the Ghidra help.

**Comment by H0r53** [comment]
@ghidracadabra the `Commit Params/Return` action within the Decompiler view worked exactly as you described. Thank you! I'm now seeing the updated function prototype when using FunctionManager.

What I'm uncertain about is how I can go about doing this programmatically. Even if I could create an updated function signature with `HighFunctionDBUtils`, I don't want to have to know the updated / inferred signature manually. Ghidra's Decompiler view seems to have a way to infer the function prototype.

How can I access the Decompiler's inferred prototype programmatically, such that I could use it to generate a new prototype via `HighFunctionDBUtils` without needing to manually review the function?

**Comment by ghidracadabra** [comment]
@H0r53 The decompiler's inferred prototype is a field of the ```HighFunction``` object.  One of the methods named ```commitParamsToDatabase``` in the ```HighFunctionDBUtils``` class takes a ```HighFunction``` as a parameter and grabs the prototype from it.

That being said, when the decompiler analyzes a function ```foo``` it considers the body of ```foo``` and any information in the program database about the prototypes of callees of ```foo```.  In particular, it does not examine what happens at call sites of ```foo```.  You might consider running the "Decompiler Parameter ID" analyzer, which will examine the entire call tree of the program, infer prototypes for each function, and commit them all to the database.  You can run this analyzer from the Code Browser via Analysis -> One Shot -> Decompiler Parameter ID (depending on the program you are analyzing this analyzer may be enabled or disabled by default).

**Comment by H0r53** [comment]
@ghidracadabra thanks again for your help. I looked into using `HighFunction` but it turns out the One Shot -> Decompiler Parameter ID analysis did exactly what I needed. Now, all of the local functions that were previously given an empty prototype have been updated in the Ghidra DB and `FunctionManager::getFunctions()` correctly returns the updated function signatures. Thank you for your patience and assistance with this. The Ghidra team here is always very helpful.
